src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- MaxPoolSingleOut.cpp - MaxPoolSingleOut Op --------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX MaxpoolSingleOut operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/IR/Value.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "llvm/ADT/ArrayRef.h"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXMaxPoolSingleOutOpLoweringToTOSA : public OpConversionPattern<ONNXMaxPoolSingleOutOp> {
+public:
+  using OpConversionPattern<ONNXMaxPoolSingleOutOp>::OpConversionPattern;
+  using OpAdaptor = typename ONNXMaxPoolSingleOutOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXMaxPoolSingleOutOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
+    Location loc = op->getLoc();
+
+    Value Input = adaptor.X();
+    // - The attribute Ceil_mode is ignored because its effect is already impacting the return
+    // type of the operator
+    // - The attribute auto_pad is ignored because it is marked as deprecated for ONNX
+    // - The attributes storage_order and dilations is also unsupported
+    mlir::StringAttr autoPad = adaptor.auto_padAttr();
+    
+    mlir::ArrayAttr dilations = adaptor.dilationsAttr();
+    mlir::ArrayAttr kernelShape = adaptor.kernel_shapeAttr();
+    mlir::ArrayAttr pads = adaptor.padsAttr();
+    mlir::ArrayAttr strides = adaptor.stridesAttr();
+
+    auto InputType = Input.getType().cast<RankedTensorType>();
+    auto resultType = op.getResult().getType();
+    auto resultShape = resultType.cast<RankedTensorType>().getShape();
+    // Construct the transposed type for the new MaxPool OP
+    Type newResultType = RankedTensorType::get(
+      {resultShape[0], resultShape[2], resultShape[3], resultShape[1]}, InputType.getElementType());
+
+    if (autoPad != "NOTSET") {
+      llvm::errs() << "auto_pad attribute is deprecated and its value will be ignored.";
+    }
+    if (dilations) {
+      llvm::errs() << "dilations attribute is unsupported by TOSA and its value will be ignored.";
+    }
+    
+    // ONNX Mlir uses NCHW as an input while TOSA expects NHWC. Insert a transpose
+    // to change the format
+    Value targetTensor = tosa::getConstTensor<int32_t>(rewriter, op, {0, 2, 3, 1}, {4}).value();
+    Type outputType = RankedTensorType::get({-1, -1, -1, -1}, InputType.getElementType());
+    Input = tosa::CreateOpAndInfer<tosa::TransposeOp>(rewriter, loc, outputType, Input, targetTensor).getResult();
```

Comment:
The conv merge introduced small helper functions for this, but it is up to you if you want to use them. They don't do anything special.